etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,187 @@
+        queue: Queue = Queue()
+        queue.put((self._hierarchy_root, 0))
+        while not queue.empty():
+            node, level = queue.get()
+            node_levels.append((level, node))
+            child_nodes = hierarchy_structure.get(node, [])
+
+            if len(child_nodes) == 0:
+                if leaves_level is not None and level != leaves_level:
+                    raise ValueError("All hierarchy tree leaves must be on the same level!")
+                else:
+                    leaves_level = level
+
+            for adj_node in child_nodes:
+                if adj_node not in seen_nodes:
+                    queue.put((adj_node, level + 1))
+                    seen_nodes.add(adj_node)
+
+        if len(seen_nodes) != num_nodes:
+            raise ValueError("Invalid tree definition: disconnected graph!")
+
+        levels = defaultdict(list)
+        for level, node in node_levels:
+            levels[level].append(node)
+
+        return levels
+
+    def _get_subtree_sizes(self, hierarchy_levels: Dict[int, List[str]]) -> Dict[str, int]:
+        """Compute subtree size for each node."""
+        subtree_size_map = dict()
+        for _, node_list in sorted(hierarchy_levels.items(), key=lambda x: x[0], reverse=True):
+            for node in node_list:
+                subtree_size = 0
+                if node not in self.level_structure:
+                    subtree_size_map[node] = 1
+                    continue
+
+                for child_node in self.level_structure.get(node, []):
+                    subtree_size += subtree_size_map[child_node]
+
+                subtree_size_map[node] = subtree_size
+
+        return subtree_size_map
+
+    def get_summing_matrix(self, target_level: str, source_level: str) -> scipy.sparse.base.spmatrix:
+        """
+        Get summing matrix for transition from source level to target level.
+
+        Parameters
+        ----------
+        target_level:
+            Name of target level.
+        source_level:
+            Name of source level.
+
+        Returns
+        -------
+        :
+            transition matrix from source level to target level
+
+        """
+        try:
+            target_idx = self._level_to_index[target_level]
+            source_idx = self._level_to_index[source_level]
+        except KeyError as e:
+            raise ValueError("Invalid level name: " + e.args[0])
+
+        if target_idx >= source_idx:
+            raise ValueError("Target level must be higher in hierarchy than source level!")
+
+        target_level_list = self.get_level_segments(target_level)
+        source_level_list = self.get_level_segments(source_level)
+        summing_matrix = lil_matrix((len(target_level_list), len(source_level_list)))
+
+        j = 0
+        for i, segment in enumerate(target_level_list):
+            sub_segment_size = self._segment_subtree_size_map[segment]
+
+            while sub_segment_size > 0:
```

Comment:
It will always be positive is distance between source and target level is more than 1

Reply:
If it remained positive, we would not be able to calculate any summing matrix with this approach.